File: src/core/utils/semver.rs
use semver::Version;
use std::collections::HashMap;
use std::str::FromStr;

pub type VersionMap = HashMap<String, String>;
pub type VersionVec = Vec<String>;
pub fn map_versions(versions: Vec<String>) -> (VersionMap, VersionVec) {
    let mut map = VersionMap::new();
    let mut vec = VersionVec::new();

    let mut major = 0;
    let mut minor = 0;
    for version in versions {
        if let Ok(version) = Version::from_str(&version) {
            if version.major != major {
                major = version.major;
                let key = major.to_string();
                map.insert(key.clone(), version.to_string());
                vec.push(key);
            }
            if version.minor != minor {
                minor = version.minor;
                let key = format!("{}.{}", major, minor);
                map.insert(key.clone(), version.to_string());
                vec.push(key);
            }
        }
    }

    vec.reverse();
    (map, vec)
}

File: src/nvm/semver.py
from semver import Version

type MapVer = dict[str, Version]
type VecVer = list[str]


def map_versions(versions: list[str]) -> tuple[MapVer, VecVer]:
    res_map: MapVer = {}
    res_vec: VecVer = []

    for version in versions:
        try:
            sv = Version.parse(version)
        except Exception:
            continue

        major = str(sv.major)
        if major not in res_map:
            res_map[major] = sv
            res_vec.append(major)
        elif sv > res_map[major]:
            res_map[major] = sv

        mm = f"{sv.major}.{sv.minor}"
        if mm not in res_map:
            res_map[mm] = sv
            res_vec.append(mm)
        elif sv > res_map[mm]:
            res_map[mm] = sv

    return res_map, res_vec[::-1]
